### ISSUE TITLE
Solve the compatibility problem between EIP-1559 and Policies

### ISSUE THREAD

**Issue by txhsl** [bane-labs/go-ethereum#187]
# Rationale

Briefly speaking, we are not fully compatible with EIP-1559 after #6 and #166. The compatibility problem between EIP-1559 and Policies are caused by a different economic model. 

- As I've mentioned in #186, EIP-1559 proposed a dynamic fee market scheme with `baseFee` and `gasTip`. Generally, when there is a network traffic, the gas price increases. But if there is no transaction in the network, `baseFee` and `gasTip` may drop to nearly 0;
- But in N3, we don't have a dynamic price model. Instead, we maintain some price factors in the Neo Policy. As a result, the system fee remains the same regardless of traffic, and the only difference is the network fee. 

There are also several facts I've noticed on testnet:

1. It is frequent that CNs create empty blocks, so the `gasPrice`/`baseFee` is pretty low;
2. To achieve a cost about 0.01 GAS per native transfer transaction (costs 21000 gas in EVM), we need a `gasPrice` about `0.01 * 10 ^ 18 / 21000 ≈ 500 Gwei`, which is incredibly higher than other EVM chains, e.g. 7 Gwei on Ethereum, 1 Gwei on BNB Smart Chain, 100 Gwei on Polygon and 0.1 Gwei on Optimistic/Arbitrum;

Empty blocks and high gas price may cause problems in EIP-1559 compatible tools, e.g. MetaMask.

We've already modified some RPC APIs to suite Neo X Policies, like changing `eth_gasPrice` and `eth_maxPriorityFeePerGas` to provide a minimum acceptable value so that a transaction will not be rejected by txpool. But `eth_feeHistory` is somehow conflict with our Policy idea, it returns a fee list of recently accepted transactions. It would be confusing when the list is empty, for EIP-1559 compatible tools, this case equals to a zero `eth_maxPriorityFeePerGas`. Their logics don't consider a minimum value, so when I increase the `minGasTipCap` limit, MetaMask hits RPC errors.

# Implementation

For the MetaMask problem, I found it can be solved by disabling `eth_feeHistory` RPC API. But this compatibility issue may happen on other tools or custom codes. Maybe they use `0.8*eth_maxPriorityFeePerGas` and hit an error.

So we may need a discussion, are we good to continue with this EIP-1559 incompatible? Or we have other ideas about the price thing?

**Comment by txhsl** [comment]
@gsmachado @chenquanyu @AnnaShaleva @roman-khimov @steven1227

**Comment by gsmachado** [comment]
This is concerning, in my opinion.

We'll certainly have cases where tools would get confused on gas estimation -- and devs would dislike it, saying "you guys promised an EVM-compatible chain, but it's not".

We need to brainstorm about how to progress on this. Maybe we should collect concrete use cases of dApps or other tools that would get lost by not sticking to EIP-1559? Then we'd be able to come up with strategies on how to overcome it.

I don't know... just brainstorming out loud.

**Comment by roman-khimov** [comment]
Glancing over https://eips.ethereum.org/EIPS/eip-1559 I'm not sure how many tools do `validate_block()`. Nodes obviously do that, but node can be fixed, otherwise `base_fee_per_gas` can be set to whatever value we need and dynamic calculation just be disabled.

The next option of course is to drop `ErrUnderpriced` and follow the EIP-1559 economics completely. But no one knows how many transactions we are going to have on mainnet. If it's around the same as for N3 then prices are gonna be close to zero and validators would get nothing. At the same time, close to zero prices may be incentivizing at least some activity. Nothing is certain here.

Then `gas_limit` (and/or `ELASTICITY_MULTIPLIER`/`BASE_FEE_MAX_CHANGE_DENOMINATOR`). These can be played to affect EIP-1559 calculations, maybe some combination can be good enough for Bane economics.

**Comment by steven1227** [comment]
I suggest change `feeHistory` rpc makes more compatible with Eip1559 and Bane economics.

**Comment by txhsl** [comment]
Considering the above comments, @chenquanyu, @steven1227 and I had a discussion. 

We decided to keep going with this N3-like pricing method, by maintaining `baseFee` and `minGasTip` in Policy contract, so that the price won't go to 0.

To avoid potential problems due to empty blocks, we decided to modify `eth_gasPrice`, `eth_maxPriorityFeePerGas` and `eth_feeHistory` RPCs to provide a default-and-suggested "price space" which should be compatible for most of the tools.
